persist/fs/seek_manager.go

```diff
@@ -137,75 +170,198 @@ func (m *seekerManager) CacheShardIndices(shards []uint32) error {
-		_, exists = byTime.seekers[tNano]
-		if exists {
-			byTime.Unlock()
-			continue
-		}
+	byTime.Lock()
+	seekersAndBloom, err := m.getOrOpenSeekersWithLock(startNano, byTime)
+	byTime.Unlock()
+	return seekersAndBloom.bloomFilter, err
+}
 
-		seeker, err := m.newOpenSeeker(byTime.shard, t)
-		if err != nil {
-			if err != errSeekerManagerFileSetNotFound {
-				// Best effort to open files, if not there don't bother opening
-				multiErr = multiErr.Add(err)
-			}
-			byTime.Unlock()
-			continue
+func (m *seekerManager) Borrow(shard uint32, start time.Time) (ConcurrentFileSetSeeker, error) {
+	byTime := m.seekersByTime(shard)
+
+	byTime.Lock()
+	// Track accessed to precache in open/close loop
+	byTime.accessed = true
+
+	startNano := xtime.ToUnixNano(start)
+	seekersAndBloom, err := m.getOrOpenSeekersWithLock(startNano, byTime)
+	if err != nil {
+		return nil, err
+	}
+
+	seekers := seekersAndBloom.seekers
+	availableSeekerIdx := -1
+	availableSeeker := borrowableSeeker{}
+	for i, seeker := range seekers {
+		if !seeker.isBorrowed {
+			availableSeekerIdx = i
+			availableSeeker = seeker
+			break
 		}
+	}
 
-		byTime.seekers[tNano] = seeker
+	// Should not occur in the case of a well-behaved caller
+	if availableSeekerIdx == -1 {
 		byTime.Unlock()
+		return nil, errNoAvailableSeekers
 	}
 
-	return multiErr.FinalError()
+	availableSeeker.isBorrowed = true
+	seekers[availableSeekerIdx] = availableSeeker
+	byTime.seekers[startNano] = seekersAndBloom
+	byTime.Unlock()
+	return availableSeeker.seeker, nil
 }
 
-func (m *seekerManager) Seeker(shard uint32, start time.Time) (FileSetSeeker, error) {
+func (m *seekerManager) Return(shard uint32, start time.Time, seeker ConcurrentFileSetSeeker) error {
 	byTime := m.seekersByTime(shard)
 
 	byTime.Lock()
-	// Track accessed to precache in open/close loop
-	byTime.accessed = true
 
 	startNano := xtime.ToUnixNano(start)
-	seeker, ok := byTime.seekers[startNano]
-	if ok {
+	seekersAndBloom, ok := byTime.seekers[startNano]
+	// Should never happen - This either means that the caller (BlockRetriever) is trying to return seekers
+	// that it never requested, OR its trying to return seekers after the openCloseLoop has already
+	// determined that they were all no longer in use and safe to close. Either way it indicates there is
+	// a bug in the code.
+	if !ok {
 		byTime.Unlock()
-		return seeker, nil
+		return errSeekersDontExist
 	}
 
-	var err error
-	seeker, err = m.newOpenSeeker(shard, start)
-	if err == nil {
-		byTime.seekers[startNano] = seeker
+	found := false
+	for i, compareSeeker := range seekersAndBloom.seekers {
+		if seeker == compareSeeker.seeker {
+			found = true
+			compareSeeker.isBorrowed = false
+			seekersAndBloom.seekers[i] = compareSeeker
+			break
+		}
 	}
+	// Should never happen with a well behaved caller. Either they are trying to return a seeker
+	// that we're not managing, or they provided the wrong shard/start.
+	if !found {
+		return errReturnedUnmanagedSeeker
```

Comment:
Need Unlock here to yeah?  Maybe we should use defer in this fn too.

Reply:
nice catch, yeah I'll convert to defer